### ISSUE TITLE
leaper/manager_42/update_crawler rewrite as origin manager

### ISSUE THREAD

**Issue by jberry-suse** [openSUSE/openSUSE-release-tools#1643]
Meta issue to collect ideas and requirements about this topic.

- [x] handle any project including `:Update` projects for lookup file
- [x] generic leaper rules that do not require tweaking for every release
- [ ] release schedule for automatic leaper rule changes and update_crawler mode changes
- [x] provide reports instead of logs
- [x] provide a web interface to leaper reports

**Comment by jberry-suse** [comment]
## clients/users of lookup file/information

- maintenance (OMG plugin) [per package]
- review [per package]
- crawler [per package]
- release manager (submission reviews, and report totals) [per package, all]

## existing workflows handled manually

- adding to SLE-workarounds project
- changing update crawler settings based on release schedule
- maintenance ignore indication in lookup file
- packages removed in origin
- packages added to an allowed origin

## goals

- incorporate manual bits of workflow
- provide summary reports at desired interval
- simplify and focus review comments
- allow for workarounds of any origin instead of just SLE
- provide abstract origin policy as configuration
- replace spaghetti code with concise policy enforcement
- replace factory-auto devel project check with origin manager (alleviate need
  for the devel project whitelist)
- provide interfaces for computing up-to-date origin information
- replace fork concept with origin workaround

## approach

- avoid storing origin information in static file (`00Meta/lookup.yml`), but
  instead derive for source comparison and request annotation
- provide CLI interface which can then be exposed via OBS Operator
- remove workarounds subproject (ex. SLE-workarounds) in favor of request
  annotation
- provide origin family expansions for use in origin configuration and schedule
- reduce policy to an abstract set of inputs that are calculated before policy
  evaluation and can be overriden via config and schedule

## policy inputs

_names are not final_

- action enum (create, update, delete)
- from_highest_priority (is the package sourced from origin of highest priority
  that contains the package)
- origin_change bool
- revert bool (old revision within order or older product family)
  - direction: forward, backward
  - forward: original from 15.0 (in 15.1) and submit from Factory where the 15.0
    revision or common point is back in Factory tree
- same_family: origin_change = true and same product family
  - submit Leap:15.0:Update to Leap:15.1 previously came from Leap:15.0
  - from Leap:15.0 to SLE-15-SP1 not same family
  - consider workarounds and desired project same family
- higher_priority bool (like SLE-workarounds to SLE proper)
  - origin_change = true and new origin is a higher priority than previous
- pending_submission bool
  - matching parallel submission to an allowed origin project
  - origin_change = true higher priority origin, but only due to pending submit
- by_maintainer bool (authored by a maintainer in origin)

## configuration

Placed in `OSRT:OriginConfig` attribute of the target project. Origin family
expansion patterns:

- `<devel>`: devel projects of target project
- `$PREFIX*`: all products with a common family prefix
- `*`: all allowed origins
- `*~`: all allowed origin workarounds
- `@$PROJECT`: inherit origins from `$PROJECT` (not schedule)

Specific projects can be indicated directly as `$PROJECT`.

Values from the `OSRT:Config` (`~/.oscrc`) may be referenced using:
`<config:$key>`.

The following optional values may be overriden, but are show with defaults.

- `additional_reviews`: `[]`
- `maintainer_review`: `true`
- `pending_submission_allow`: `false`
- `automatic_updates`: `true`

An example for openSUSE:Leap:15.0:

```yaml
origins:
  <devel>:
    additional_reviews:
    - <config:review-team>
    - <config:legal-review-group>
  SUSE:SLE-15*: {} # Alternatively, <family:SUSE:SLE-15>
  openSUSE:Factory: {}
  *~: # alternatively: *(workarounds) or SUSE:SLE-15*(workarounds)
    additional_reviews:
    - <config:review-team>
    - <config:legal-review-group>

schedule:
  # no automatic factory updates, manual approval
  2018-04-24:
    openSUSE:Factory:
      automatic_updates: false
    <devel>:
      automatic_updates: false
  # no automatic SLE updates, manual approval, but allow emergency in parallel
  2018-05-14:
    SUSE:SLE-15*:
    - automatic_updates: false
    - pending_submission_allow: true
```

The overrides in the schedule are applied cumulatively on and beyond the date
indicated in the schedule.

An example of a maintenance project `openSUSE:Leap:15.1:Update`:

```yaml
origins:
# inheritted from parent project: openSUSE:Leap:15.1
- @openSUSE:Leap:15.1
# could override or add origins
```

An example of a SLE project `SUSE:SLE-15-SP1:GA`:

```yaml
origins:
  *:
    maintainer_review: false
  <devel>: {}
  SUSE:SLE-15*: {}
  openSUSE:Factory:
  - pending_submission_allow: true
```

An example of `openSUSE:Factory`:

```yaml
origins:
- <devel>
```

## policy psuedocode

_The following is rough pseduocode, and not final._ The general concept sticks
with always accepting or ignoring and never declining. Ignore requests that are
dependent on pending state unless allowed via config (pending_submission_allow).

When reviews are added the message will indicate why the review was added. The
review comment itself will be kept minimal and focused.

```python
policy_apply():
  policy_ret = policy()
  if policy_ret is True:
    if maintainer_review and maintainer_review_requested:
      additional_reviews.add("<by_package>") # origin package

    reviews_add(additional_reviews)

  comment_write()
  return policy_ret

policy():
  ret = True
  if not pending_submission_allow and pending_submission:
    ret = Ignore

  if new_package:
    if from_highest_priority:
      if by_maintainer:
        return ret
      else:
        return ret, but request maintainer review
    else:
      if by_maintainer:
        return ret, but request release manager review
      else:
        return ret, but request maintainer and release manager review
  elif delete_package:
    if origin_change: # meaning still present in origin
      return ret, but request release manager review

    return ret
  else:
    if origin_change:
      if higher_priority:
        if same_family:
          return ret

        return ret, release manager
      else:
        return ret, release manager (going backwards comment)
    else:
      if revert:
        return ret, but request release manager review

      if automatic_updates:
        return ret
      else:
        return ret, but request release manager review
```

## review comment

The comment should be kept as minimal as possible to aid in speedy decision by
release managers. Worth considering keeping the bullet format even when origin
is not changing (or new) just so consistent location. Focus on report content
and style as formatting can be changed easily.

Only include target information if multi-action submission (maintenance).

The following are some 

- new package from Factory by maintainer

  - Matched origin: openSUSE:Factory

- also available in SUSE:SLE-15-SP1:GA, SUSE:SLE-15:Update, SUSE:SLE-15:GA

  - Matched origin: openSUSE:Factory
  - Suggested origin: SUSE:SLE-15-SP1:GA

- update from SUSE:SLE-15:GA to SUSE:SLE-15-SP1:GA

  Origin changing to higher priority one within the same product family:

  - SUSE:SLE-15:GA
  - SUSE:SLE-15-SP1:GA

- update from SUSE:SLE-15-SP1:GA to SUSE:SLE-15:Update

  Origin changing to lower priority one within the same product family:

  - SUSE:SLE-15-SP1:GA
  - SUSE:SLE-15:Update

- update from SUSE:SLE-15:Update to SUSE:SLE-15:Update

  Origin changing to higher priority one within the same product family:

  - SUSE:SLE-15-SP1:GA
  - SUSE:SLE-15:Update

- update from SUSE:SLE-15:GA to openSUSE:Factory

  Origin changing to lower priority one in a different product family:

  - SUSE:SLE-15:GA (current)
  - openSUSE:Factory (request)

- update from openSUSE:Factory to SUSE:SLE-15:GA

  Origin changing to higher priority one in a different product family:

  - openSUSE:Factory (current)
  - SUSE:SLE-15:GA (request)

- update to openSUSE:Factory package which is pending

  Origin unchanged: openSUSE:Factory

  Pending request#1234 which must be accepted first

- update to openSUSE:Factory package

  Origin unchanged: openSUSE:Factory

- update to SUSE:SLE-15-SP1:GA package which is pending

  Origin unchanged: SUSE:SLE-15-SP1:GA

  Pending request#1234, but parallel submission allowed

- update to SUSE:SLE-15-SP1:GA package as workaround

  Origin changing to workaround:

  - SUSE:SLE-15-SP1:GA
  - SUSE:SLE-15-SP1:GA~

- update to openSUSE:Factory package as workaround

  Origin changing to workaround

  - openSUSE:Factory
  - openSUSE:Factory~

- change from openSUSE:Factory to devel

  change_devel request which will match once idental sources are available

- update to btrfsprogs package as workaround

  Origin changing to workaround:

  - filesystems
  - filesystems~

- update SUSE:SLE-15-SP1:GA package that is already a fork of openSUSE:Factory

  Origin unchanged: openSUSE:Factory~

- update SUSE:SLE-15-SP1:GA package that is fork of openSUSE:Factory to non-fork

  Origin dropping workaround:

  - openSUSE:Factory~
  - openSUSE:Factory

- update to SUSE:SLE-15-SP1:GA package which has a different pending submission

  Origin changing to workaround:

  - SUSE:SLE-15-SP1:GA
  - SUSE:SLE-15-SP1:GA~

  Pending request#1234, but sources do not match

- revert to openSUSE:Factory package using prior revision (self submission)

  Origin unchanged: openSUSE:Factory


## request annotation

Request comment psuedometa will include the outcome origin at the time request
was reviewed. If the request is accepted the comment will act as an _annotation_
to be used if deleted from origin or to indicate what the desired origin is of a
workaround (source not found in any allowed origin).

Instead of submitting to `SLE-workarounds` the act of accepting the request
denotes the package as sourced as a workaround.

## query interfaces

A new `osc` subcommand will be provided and exposed by OBS Operator server.

```
osc origin -p openSUSE:Leap:15.1 package Mesa
X11:XOrg
```

```
osc origin -p openSUSE:Leap:15.1 package Mesa
X11:XOrg~
```

```
https://operator.opensuse.org/origin/package/openSUSE:Leap:15.1/Mesa
X11:XOrg
```

```
osc origin -p openSUSE:Leap:15.1 list
Mesa    X11:XOrg
libclc  X11:XOrg
...
```

```
https://operator.opensuse.org/origin/list/openSUSE:Leap:15.1
Mesa    X11:XOrg
libclc  X11:XOrg (behind by 7 revisions)
...
```

```
osc origin -p openSUSE:Leap:15.1 report
Kernel:openSUSE-15.1:Submit   10   0%
X11:XOrg                      10   0%
openSUSE:Factory            8000  70%
SUSE:SLE-15-SP1:GA          1500  15%
SUSE:SLE-15:Update           300   3%
SUSE:SLE-15:GA               500   5%
...
```

```
https://operator.opensuse.org/origin/report/openSUSE:Leap:15.1
...
```

The report could be scheduled to send out e-mails on weekly basis or somesuch.

## crawler

The update crawler will be updated to respect the schedule, handle deletes, and
if desired include the fcc_submitter functionality. If desired the crawler can
be built to listen for package submission events and backfill using recent
changes.

Short of the config being changed after a request is created by the crawler and
reviewed the crawler should never make requests that are undesirable based on
the policy.

**Comment by coolo** [comment]
Your config proposal looks complete, but I dislike one aspect heavily: You rely on request comments as source of information and we really should stop doing that. This information should be present on a review web app and stored in there.

OBS comments should only be used as notifications in case of a problem. Your suggested format for the comment is targeted again towards the release manager - but it's sent to submitters as well and is in no way meaningful to them. Hence the requirement to have these reports (in case of a problem) pointing to a web page with more informations.

**Comment by lnussel** [comment]
certainly the right direction, I'd probably drop the inheritance in favor of a copy for simplicity. I also have doubts about dropping the lookup file entirely in favor of the  operator server part.  Looks like the operator server part can replace the package diff app I have on tanana though. nice :-)
Wrt comment I guess that boils down more or less to an abstract reporting and control interface the bots need to get. So far we didn't distinguish interfaces for different roles.

How to handle the origin change ACK? no more ATleaper override accept?

**Comment by coolo** [comment]
I don't mind if we drop the lookup file if there is an easy enough URL to point e.g. screening team at.

That can be https://operator.opensuse.org/origin/list/openSUSE:Leap:15.1 as referenced - if it's quick to load.

**Comment by jberry-suse** [comment]
Summary of updates after discussion with @lnussel:

- bring back fork concept for packages to never come from anywhere else (like SLE-only)
- new packages focused to introduced in SLE / openSUSE maintenance updates (fcc process out of scope)
- agreed, deletes dropping to next available origin makes sense and automates updates to SLE that are later removed after maintenance update
  - if after deletion it a package no longer patches any other origin, but is present indicate may want to resubmit from that origin (and provide action button)


open points

- whether to block bot review until release manager review is accept to allow for changes that may alleviate it (useful for factory case with mass submission from new devel)
  - practically may be more trouble than worth and specifically in devel case would have no way to close the release manager group review unless the bot was also a member
- need to handle packages that originate from Factory, but are not to be updated since they came from previous leap version (mostly for crawler) and related batch updating such packages when desired
  - perhaps some sort of _paused_ concept

points of clarification

- no more need to "override" review unless breaking policy like accepting an update that is still pending in Factory in Leap when not allowed. instead accepting the review indicate the decision

**Comment by jberry-suse** [comment]
As for depending on comments for request annotation concept, it can certainly be done in a variety of means, but the issue is that OBS is the source of truth since it stores the sources which indicate what origin things were from. Also stores the requests that document how it got that way. The lookup file as it stands is essentially a _lazy_ or cached version of the information that is only updated when needed.

The proposal suggests calculating the information always which removes the _lazy update_ part which from discussion @lnussel agrees seems best, but we'll have to see how it pans out. Either way what we end up storing if to avoid using comments is the lookup file. Storing a key/value set of string->string (package to origin) in a an external database seems like some serious over-engineering and does not offer any real benefits. A web interface can be created without a database and simply wrap OBS (just like OBS Operator).

The Origin Manager is very much a tool tied to OBS and not a separate workflow. If metadata could be added to submit/delete requests that could alternatively be used.

The only reason the annotation is needed indicate something is a workaround. In all othercases the source md5 should be found in an origin and as such that is the only storage needed. For deletes this is useful to provide additional context, but not required to meet requirements.

I intend to implement all actions via OBS Operator so that they can make use of our existing code, share code with review portion, and be scripted on command line for odd bulk operations. As such the web interface becomes a rather simple html wrapper that provides links and buttons. This was also the reason I suggested it could just be done as javascript on top of OBS either hosted by them or via userscript. Then the buttons show up were you want them instead of on a separate page.

If we really want to have separate storage it can be done, but the cons outweigh the pros in my estimation and is not necessary for a web interface.

**Comment by coolo** [comment]
As I said: I see 2 requirements that seem a little undervalued in your proposal
- A quick link to check where 15.1 packages are coming from. I don't care too much how it's stored as long as it's quick
- Comments/Notifications should only happen if there is a problem - approvals should link in a check/review to the reasoning without notification.

I never asked for an external database, but I expect it to be required for quick operations. But more as a cache. Ludwig's code stream diff application requires quite some background jobs.

**Comment by coolo** [comment]
But combining javascript with osc code sounds like an interesting strategy.

**Comment by jberry-suse** [comment]
> * A quick link to check where 15.1 packages are coming from. I don't care too much how it's stored as long as it's quick

This was in original proposal: https://operator.opensuse.org/origin/package/openSUSE:Leap:15.1/Mesa with CLI equivalent. Should be ~3 api calls behinds the scenes so no worse than anything else we have.

> * Comments/Notifications should only happen if there is a problem - approvals should link in a check/review to the reasoning without notification.

The annotation is only useful when origin is changing which is when a comment would be made and approval requested.

**Comment by coolo** [comment]
> > * A quick link to check where 15.1 packages are coming from. I don't care too much how it's stored as long as it's quick
> 
> This was in original proposal: https://operator.opensuse.org/origin/package/openSUSE:Leap:15.1/Mesa with CLI equivalent. Should be ~3 api calls behinds the scenes so no worse than anything else we have.

This requires to type the package name as URL. That's not a 'quick link'. That's why I repeated the requirement.

> 
> > * Comments/Notifications should only happen if there is a problem - approvals should link in a check/review to the reasoning without notification.
> 
> The annotation is only useful when origin is changing which is when a comment would be made and approval requested.

That's the 'should only happen if' part. I'm saying also approvals should have a link in the review message / status check. Could be e.g. https://ibs-operator.suse.de/origin/request/29292

**Comment by lnussel** [comment]
harder than on ours?

**Comment by coolo** [comment]
equally hard I'm afraid. which makes waiting for them unfeasible :)

**Comment by jberry-suse** [comment]
> This requires to type the package name as URL. That's not a 'quick link'. That's why I repeated the requirement.

So what then? Having a link on the request is something that would be possible with the userscript approach I suggested. The comment itself includes the information (ie. origin), but you only want comment when changing...

The point is the information is easily accessible. Where and how a link is provided there are many options. The most intuitive is userscript.

> That's the 'should only happen if' part. I'm saying also approvals should have a link in the review message / status check. Could be e.g. https://ibs-operator.suse.de/origin/request/29292

Link and annotation are two entirely different things. Can always have a link, but the annotation is only useful when changing. If status report API for requests is usable can place a link there always...otherwise put it in a comment...I am really not sure where you are going with this.

**Comment by coolo** [comment]
> So what then? Having a link on the request is something that would be possible with the userscript approach I suggested. The comment itself includes the information (ie. origin), but you only want comment when changing...

The link is not about requests - the link is about checking the facts at any time. Like a bookmark for the screening team or the security team or ...
 
> > That's the 'should only happen if' part. I'm saying also approvals should have a link in the review message / status check. Could be e.g. https://ibs-operator.suse.de/origin/request/29292
> 
> Link and annotation are two entirely different things. Can always have a link, but the annotation is only useful when changing. If status report API for requests is usable can place a link there always...otherwise put it in a comment...I am really not sure where you are going with this.

The review message is text - so you can put a link there as well. In the current UI it's not rendered as link, but it still belongs there. Status report API is supposed to replace it 'soonish' (in obs terms) though.

**Comment by jberry-suse** [comment]
> The link is not about requests - the link is about checking the facts at any time.

In comparison to lookup file I still have to either type the package name via ctrl+f or squint until I find the one.

**Comment by coolo** [comment]
> > The link is not about requests - the link is about checking the facts at any time.
> 
> In comparison to lookup file I still have to either type the package name via ctrl+f or squint until I find the one.

Right. What's the question?

**Comment by jberry-suse** [comment]
> Right. What's the question?

Where lies the problem?
